multiprocess: Add bitcoin-gui -ipcconnect option

ryanofsky opened the pull request:
<!-- begin based-on -->
**This is based on #29409 + # + #10102 + #19460.** The non-base commits are:

- [`bb4f280f349` multiprocess: Add bitcoin-gui -ipcconnect option](https://github.com/bitcoin/bitcoin/pull/19461/commits/bb4f280f34902a41846463fdc41721c8f7382c79)<!-- end -->

---

Building on #10102, this adds an `-ipcconnect` option to `bitcoin-gui` that connects the GUI to an existing `bitcoin-node` process already running in the background instead of spawning a new `bitcoin-node` process. This allows the GUI to be started and stopped independently of the node. By default with this change, `bitcoin-gui` will check if a `<datadir>/sockets/node.sock` socket exists and try to connect to that. If that doesn't work, it will spawn a new node process and start up the same way it did before this PR.

The default `bitcoin-gui` connect option is `-ipcconnect=auto`, which tries to connect if possible as described above, and spawns a new `bitcoin-node` process if not possible. Other supported options are `-noipcconnect` to never connect to an existing node and always spawn a new one, `-ipcconnect` to require a connection and fail if it can't be established, and `-ipcconnect=unix:<socket>` to require a connection and use a custom socket path.

With this PR, basic functionality works and gui instances can connect and disconnect from a running node. But there are rough edges: If a gui process doesn't shut down cleanly, the node can see unhandled IpcExceptions, and if node command line options are passed to `bitcoin-gui` and `bitcoin-gui` connects to an exiting bitcoin-node process instead of spawning a new one, the node options will be silently ignored. 

These changes require multiprocess support and this PR has no effect unless bitcoin is configured with `--enable-multiprocess` as described in [doc/multiprocess.md](https://github.com/bitcoin/bitcoin/blob/master/doc/multiprocess.md)

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
doc: Multiprocess misc doc and comment updates

ryanofsky committed:
multiprocess: Add capnp wrapper for Node interface

ryanofsky committed:
multiprocess: Expose Chain interface

Expose Chain interface to external processes spawning or connecting to
bitcoin-node.

ryanofsky committed:
multiprocess: Add capnp serialization code for bitcoin types

ryanofsky committed:
multiprocess: Make bitcoin-node spawn a bitcoin-wallet process

Spawn wallet subprocess instead of running wallet code internally

ryanofsky committed:
Add capnp wrapper for Chain interface

ryanofsky committed:
Add capnp wrapper for Handler interface

ryanofsky committed:
combine_logs: Handle multiprocess wallet log files

ryanofsky committed:
multiprocess: Make bitcoin-gui spawn a bitcoin-node process

Spawn node subprocess instead of running node code internally

ryanofsky committed:
multiprocess: Add bitcoin-gui -ipcconnect option

Add `-ipcconnect` option to `bitcoin-gui` to allow connecting gui to an
existing node instead of starting a new node process, so it is possible to
start and stop the gui independently of the node.

This change doesn't add an -ipcbind option to the bitcoin-wallet (which would
allow the gui to control wallet processes without going through the node), but
this would be a logical extension to add in the future.

laanwj commented:
Concept ACK, I've only reviewed the option help yet but `-ipcconnect` SGTM as option name, it reminds of `-rpcconnect` of the `-cli` client.

laanwj commented on src/bitcoin-wallet.cpp:
```diff
@@ -28,15 +30,18 @@ static void SetupWalletToolArgs()
     gArgs.AddArg("-debug=<category>", "Output debugging information (default: 0).", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     gArgs.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -debug is true, 0 otherwise).", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    if (include_ipc) {
+        gArgs.AddArg("-ipcconnect=<address>, -noipcconnect", "Connect to bitcoin-node process in the background to perform online operations. Valid <address> values are 'auto' to try connecting to default socket in <datadir>/sockets/node.sock, but proceed offline if it isn't available, 'unix' to connect to the default socket and fail if it isn't available, 'unix:<socket path>' to connect to a soc ...
```
I think this is the only case of multiple options separated by `,` inside the first argument of `ArgsManager::AddArg`. I don't know if it's strictly wrong, as it's not currently parsed in any way, but maybe the idea is to do so in the future. @MarcoFalke?

MarcoFalke replied:
No plans right now to touch the command line help in any way, so this might be fine as is

ryanofsky reviewed (commented):
Rebased 70aaca9ebecb802f7da4babd27a79a0551225475 -> 4041990fcba01a5c6eb22b827b2220f2bd8ee666 ([`pr/ipc-gui.1`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.1) -> [`pr/ipc-gui.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.1-rebase..pr/ipc-gui.2)) on top of #19460 pr/ipc-connect.3
Rebased 4041990fcba01a5c6eb22b827b2220f2bd8ee666 -> 0714770d6bf52fbf46a50aff298df8ffdf4eb3ff ([`pr/ipc-gui.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.2) -> [`pr/ipc-gui.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.2-rebase..pr/ipc-gui.3)) on top of #19460 pr/ipc-connect.4
Rebased 878e11049c97564f81018381b88bf314ce4c4f01 -> 5c05c38c1a6185ee18eaf74948cf28fd1560abf4 ([`pr/ipc-gui.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.4) -> [`pr/ipc-gui.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.4-rebase..pr/ipc-gui.5)) on top of #19460 pr/ipc-connect.7
Rebased 5c05c38c1a6185ee18eaf74948cf28fd1560abf4 -> 2b9fbc08cc9f27dd092798b75afa1e2318e749c9 ([`pr/ipc-gui.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.5) -> [`pr/ipc-gui.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.5-rebase..pr/ipc-gui.6)) on top of #19460 pr/ipc-connect.8
Rebased 2b9fbc08cc9f27dd092798b75afa1e2318e749c9 -> 6f39b4689aea86283b43add4bdebda11dee4d73e ([`pr/ipc-gui.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.6) -> [`pr/ipc-gui.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.6-rebase..pr/ipc-gui.7)) on top of #19460 pr/ipc-connect.9
Rebased 6f39b4689aea86283b43add4bdebda11dee4d73e -> e2e2a06bf91cc321eb7e6cab2c19cfc4984dfe1c ([`pr/ipc-gui.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.7) -> [`pr/ipc-gui.8`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.7-rebase..pr/ipc-gui.8)) on top of #19460 pr/ipc-connect.10
Rebased e2e2a06bf91cc321eb7e6cab2c19cfc4984dfe1c -> 137225d7ba437d7d14e122a3eb35512ef8648041 ([`pr/ipc-gui.8`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.8) -> [`pr/ipc-gui.9`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.8-rebase..pr/ipc-gui.9)) on top of #19460 pr/ipc-connect.14
Rebased 137225d7ba437d7d14e122a3eb35512ef8648041 -> 871891d38cb3a4b73fa2016de70990f221662d90 ([`pr/ipc-gui.9`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.9) -> [`pr/ipc-gui.10`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.9-rebase..pr/ipc-gui.10)) on top of #19460 pr/ipc-connect.15
Rebased 871891d38cb3a4b73fa2016de70990f221662d90 -> 4b045d994b82bb24d0cab1efd5488575b9e7851d ([`pr/ipc-gui.10`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.10) -> [`pr/ipc-gui.11`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.10-rebase..pr/ipc-gui.11)) on top of #19460 pr/ipc-connect.16 due to conflict with bitcoin-core/gui#233
Rebased 4b045d994b82bb24d0cab1efd5488575b9e7851d -> 7cb12b4bc81c7e8978830e427f9d36b8590d9848 ([`pr/ipc-gui.11`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.11) -> [`pr/ipc-gui.12`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.11-rebase..pr/ipc-gui.12)) on top of #19460 pr/ipc-connect.19
Rebased 7cb12b4bc81c7e8978830e427f9d36b8590d9848 -> e7dfe88e2ab772c4eed939dd5bda53c42827cac8 ([`pr/ipc-gui.12`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.12) -> [`pr/ipc-gui.13`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.12-rebase..pr/ipc-gui.13)) on top of #19460 pr/ipc-connect.21
Rebased e7dfe88e2ab772c4eed939dd5bda53c42827cac8 -> 4dc8af0440073813101b85360482713c7abc6949 ([`pr/ipc-gui.13`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.13) -> [`pr/ipc-gui.14`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.13-rebase..pr/ipc-gui.14)) on top of #19460 pr/ipc-connect.22
Rebased 4dc8af0440073813101b85360482713c7abc6949 -> c997272eb0a6c805f098c920d8210221f770672f ([`pr/ipc-gui.14`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.14) -> [`pr/ipc-gui.15`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.14-rebase..pr/ipc-gui.15)) on top of #19460 pr/ipc-connect.24 due to conflict with bitcoin-core/gui#381
Rebased c997272eb0a6c805f098c920d8210221f770672f -> e23cf8171fdd6e1187731dd2393fcd64469ffd80 ([`pr/ipc-gui.15`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.15) -> [`pr/ipc-gui.16`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.15-rebase..pr/ipc-gui.16)) on top of #19460 pr/ipc-connect.25 due to conflict with bitcoin-core/gui#434
Rebased e23cf8171fdd6e1187731dd2393fcd64469ffd80 -> f2f33658dddd45477e43bcad749833f6fd336bb2 ([`pr/ipc-gui.16`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.16) -> [`pr/ipc-gui.17`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.16-rebase..pr/ipc-gui.17)) on top of #19460 pr/ipc-connect.26
Rebased f2f33658dddd45477e43bcad749833f6fd336bb2 -> f8d4a15805296edb365602a37875637610fe77c7 ([`pr/ipc-gui.17`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.17) -> [`pr/ipc-gui.18`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.17-rebase..pr/ipc-gui.18)) on top of #19460 pr/ipc-connect.28

meshcollider commented:
Concept ACK

Sjors commented:
Needs rebase 🎉

What was the magic incantation again to tunnel a unix socket over SSH? I'll try if I can run the GUI on a different machine than the node.

ryanofsky committed:
multiprocess: Add debug.log .wallet/.gui suffixes

Add .wallet/.gui suffixes to log files created by bitcoin-gui and
bitcoin-wallet processes so they don't clash with bitcoin-node log file.

ClaraBara22 reviewed (approved):
Thank you

ryanofsky committed:
util: Add util::Result workaround to be compatible with libmultiprocess

Make default constructor more generic so it doesn't only work with void types.

Sjors commented:
> What was the magic incantation again to tunnel a unix socket over SSH? I'll try if I can run the GUI on a different machine than the node.

On an Ubuntu 22.10 machine I ran `src/bitcoin-node -ipcbind=unix:///tmp/node.sock`. I then opened an SSH connection: `ssh -L/tmp/remote-node.sock:/tmp/node.sock ubuntu`. And then on the mac I connected the GUI: `src/bitcoin-gui -ipconnect=unix:///tmp/remote-node.sock`


I connected just after  the node caught up on old blocks. It had the most recent block, but the logs show it as 0.999998% finished. The UI displayed a nonsensical status for that:
<img width="665" alt="Scherm­afbeelding 2023-02-11 om 12 33 33" src="https://user-images.githubusercontent.com/10217/218255851-ca24f92a-fd1a-4e41-a9a8-0f7c541c29aa.png">

Once a new block came in a few minutes later, the sync dialog was correctly dismissed. I haven't tried reproducing this without the IPC stuff, so it might be unrelated.

Other than that, it seems to work. Cool stuff!

Sjors commented:
> When I exit the GUI and after that stop `bitcoind` the latter crashes.

This still happens on 5f91aa3d8b47cee97e625c3b5689a37dbd941eff. Same setup as above: node and wallet running on Ubuntu 23.04 with a GUI client connecting from macOS using a unix socket of SSH.

But now I've notice this happens even if you don't connect to it, i.e. starting without `-ipcbind`.

After letting the chain sync (a few days worth of blocks), I shut it down. At that point the log (on Ubuntu):

```
2023-05-05T10:30:24Z UpdateTip: new best=000000000000000000010ed2e719320bb8bdea58868cf0bee4897b33b7c20c14 height=788138 version=0x32b96000 log2_work=94.158122 tx=831696670 date='2023-05-03T20:58:19Z' progress=0.999404 cache=515.2MiB(4072240txo)
2023-05-05T10:30:29Z Synchronizing blockheaders, height: 788359 (~100.00%)
^C2023-05-05T10:30:42Z [rpc] Interrupting HTTP RPC server
2023-05-05T10:30:42Z [rpc] Interrupting RPC
2023-05-05T10:30:42Z tor: Thread interrupt
2023-05-05T10:30:42Z torcontrol thread exit
2023-05-05T10:30:42Z addcon thread exit
2023-05-05T10:30:42Z Shutdown: In progress...
2023-05-05T10:30:42Z opencon thread exit
2023-05-05T10:30:42Z [rpc] Stopping HTTP RPC server
2023-05-05T10:30:42Z [rpc] Stopping RPC
2023-05-05T10:30:42Z [rpc] RPC stopped.
terminate called after throwing an instance of 'ipc::Exception'
  what():  kj::Exception: kj/async-io-unix.c++:498: disconnected: ::read(fd, buffer, maxBytes): Connection reset by peer
stack: 7fb82ba25644 7fb82ba34eab 7fb82b9c41b5 7fb82ba1bdb0 7fb82bb95720 7fb82bb8ec20 7fb82bb92df0 7fb82bbf6b30 7fb82bbf1220 7fb82bbe6cc0 7fb82bbe2800 7fb82bbd25d0 7fb82bbd3084 7fb82bbcf610 564eea2664d0 564eea26ad60
Aborted (core dumped)
```

I then have to restart the node and sync again.

Sjors commented:
I'm able to produce another crash that may be related to this PR. When using an external signer that's connected to the Ubuntu machine, and trying to spend with it from the macOs GUI. After I click Sign on device, and hit "send":

```
Assertion failed: (!complete), function sendButtonClicked, file sendcoinsdialog.cpp, line 525.
zsh: abort      src/bitcoin-gui -ipcconnect=unix:///tmp/remote-node.sock
```

It also crashes if I create a PSBT (options -> wallet -> enable psbt controls) instead, just a different line

```
Assertion failed: (!complete), function sendButtonClicked, file sendcoinsdialog.cpp, line 509.
```

Both nodes were configured with have `--enable-external-signer`. I'm able to call `enumeratesigners` from the console.

These asserts happen after calling `fillPSBT(` _without_ signing (This prevents an external from being called prematurely). `complete` can't be possibly be true, hence the assert.

ryanofsky commented:
Thanks Sjors, I'm trying to reproduce the shutdown error but didn't succeed yet. Will try again with a mainnet node. I didn't look into the external signer bug yet, but maybe that is more straightforward. In both cases running with -debug=ipc could make it clearer what's going on, especially in the shutdown case. Probably what is happening in that case is that some thread is trying make an IPC call after the socket is closed, so an ipc::Exception is raised which is uncaught. Fix might be synchronizing the IPC call & socket close, or just catching the exception and handling it

ryanofsky committed:
multiprocess: Add capnp wrapper for Wallet interface

ryanofsky committed:
test: Fix multiprocess test for unclean shutdown on kill

ryanofsky committed:
Add capnp serialization code for bitcoin types

- Add capnp ToBlob, ToArray, Wrap, Serialize, and Unserialize helper functions
- Add support for std::chrono::seconds capnp serialization
- Add support for util::Result capnp serialization

ryanofsky committed:
multiprocess: Add bitcoin-wallet -ipcconnect option

Add `-ipcconnect` option to `bitcoin-wallet` to allow connecting to a bitcoin
node process over IPC.  The `bitcoin-wallet` tool doesn't really do anything with its
connection to the node yet, but it could potentially run or serve RPCs that
require being online.

Example usage:

    src/bitcoin-node -regtest -debug -ipcbind=unix
    src/bitcoin-wallet -regtest -ipcconnect=unix info

jimhashhq commented:
As promised in #10102 comments, I tried out `-ipcconnect` using rebased pr-19641 (thank you); everything seemed to work consistent with my new understanding of multiprocess interactions (thanks again).  I did notice one issue:

1) If `bitcoin-node` is run with -`disablewallet`. A subsequent start of a gui with`-ipcconnect=auto`, but _without_ `-disablewallet` will crash `bitcoin-node`.  This is obviously an operator error, but may be worth handling w/o node crash. Error is:

> 23329 Segmentation fault (core dumped) bitcoin-gui -conf=regtest-guisolo.conf -ipcconnect=auto -debuglogfile=bitcoin-gui-`date +"%Y-%m-%d"`.log 2>&1

Note: `bitcoin-gui`and `bitcoin-node` work fine if `bitcoin-gui` correctly started with `-disablewallet`.
Also, `bitcoin-cli` wallet commands run against a `bitcoin-node` with `-disablewallet` (again, an operator error) correctly just return a `Method not found` RPC error and do not cause node crash.

Please reference `-disablewallet` notes at [docs/developer-notes.md#wallet](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#wallet)

Interestingly, a _prior revision_ to [docs/developer-notes.md#wallet](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#wallet) expands on this and references an obsolete/superceded test:
```
   - *Rationale*: In RPC code that conditionally use the wallet (such as
     `validateaddress`) it is easy to forget that global pointer `pwalletMain`
     can be NULL. See `qa/rpc-tests/disablewallet.py` for functional tests
     exercising the API with `-disablewallet`
```
I did not yet try and debug the crash as I'm not totally sure I am working on the correct integration branch for this?
Hopefully some of this is useful, thank you.
See also:  #19460

jimhashhq commented:
> I created https://github.com/bitcoin-core/libmultiprocess/issues/169 to track it.

Thank you; glad to hear you were able to easily reproduce.  I did not report an issue myself only because:
1) Didn't want to sound alarmist; this is isolated to the experimental multiprocess feature.
2) Calling this just an issue might avoid addressing deeper process control hierarchy challenges,
prioritizing a temporary implementation over longer term design (below).
3) I'm not currently qualified to opine on any potential pitfalls of mixing/matching legacy Boost 
threading/synchronization with KJ cooperative multitasking, coroutines, etc.
> [Comment by @ryanofsky bitcoin-core/libmultiprocess pr-169](https://github.com/bitcoin-core/libmultiprocess/issues/169#issuecomment-2783923191):
> I think ideal fix in long run would be for bitcoin-wallet processes to handle wallet RPC without relying on bitcoin-node, and for bitcoin-gui to communicate with bitcoin-wallet processes directly without going through bitcoin-node. At that point, the bitcoin-node -disablewallet option could be dropped.

Saw _one_ other statement to this effect, but can't place it RN, will reference here if I can.

I also would also be interested in seeing multiprocess achieve the goal of *not* having the node act as a go-between for wallet RPCs, and might be personally interested on working on any formal software "interface rationalization" spec to that end, if one doesn't exist already.

ryanofsky committed:
test: Increase feature_block.py and feature_taproot.py timeouts

Needed because BlockConnected notifications are a lot slower with the wallet
running in separate process.

ryanofsky commented:
<!-- begin push-31 -->
Rebased dbced82bbafd8e4cc62dc5d91a6531736dff2caf -> f956e90fbdeff8c184ea2e6fb9362b2d3f12f30f ([`pr/ipc-gui.30`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.30) -> [`pr/ipc-gui.31`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.31), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.30-rebase..pr/ipc-gui.31))<!-- end -->
<!-- begin push-32 -->
Rebased f956e90fbdeff8c184ea2e6fb9362b2d3f12f30f -> 6bab3c091b6e9f8a90255066e7fb7f063eede5bc ([`pr/ipc-gui.31`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.31) -> [`pr/ipc-gui.32`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.32), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.31-rebase..pr/ipc-gui.32))<!-- end -->
<!-- begin push-33 -->
Rebased 6bab3c091b6e9f8a90255066e7fb7f063eede5bc -> 8e4ceb66a97fa14afc932d12f944662638ed87af ([`pr/ipc-gui.32`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.32) -> [`pr/ipc-gui.33`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.33), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.32-rebase..pr/ipc-gui.33))<!-- end --> on top of #19460 pr/ipc-connect.48
<!-- begin push-34 -->
Rebased 8e4ceb66a97fa14afc932d12f944662638ed87af -> 1b7e9b30bdb3db31ee94ed500c7437eaba817581 ([`pr/ipc-gui.33`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.33) -> [`pr/ipc-gui.34`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.34), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.33-rebase..pr/ipc-gui.34))<!-- end -->
<!-- begin push-35 -->
Rebased 1b7e9b30bdb3db31ee94ed500c7437eaba817581 -> 7f80749c5140dea02c411931e5df165b4cf7aec5 ([`pr/ipc-gui.34`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.34) -> [`pr/ipc-gui.35`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.34-rebase..pr/ipc-gui.35))<!-- end -->
<!-- begin push-36 -->
Rebased 7f80749c5140dea02c411931e5df165b4cf7aec5 -> 86eb826842691ed52057a1aada702a5c1eb7c262 ([`pr/ipc-gui.35`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.35) -> [`pr/ipc-gui.36`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.36), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.35-rebase..pr/ipc-gui.36))<!-- end --> on top of #19460 pr/ipc-connect.51
<!-- begin push-37 -->
Rebased 86eb826842691ed52057a1aada702a5c1eb7c262 -> 93a1d8e9fb0adb578656a42b30aeb821ba193f34 ([`pr/ipc-gui.36`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.36) -> [`pr/ipc-gui.37`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.37), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.36-rebase..pr/ipc-gui.37))<!-- end --> on top of #19460 pr/ipc-connect.52
<!-- begin push-38 -->
Rebased 93a1d8e9fb0adb578656a42b30aeb821ba193f34 -> bb4f280f34902a41846463fdc41721c8f7382c79 ([`pr/ipc-gui.37`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.37) -> [`pr/ipc-gui.38`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.38), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.37-rebase..pr/ipc-gui.38))<!-- end --> on top of #19460 pr/ipc-connect.53
<!-- begin push-39 -->
Rebased bb4f280f34902a41846463fdc41721c8f7382c79 -> 5f93661901bcaee4731a02b86a31712f237629af ([`pr/ipc-gui.38`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.38) -> [`pr/ipc-gui.39`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-gui.39), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-gui.38-rebase..pr/ipc-gui.39))<!-- end --> on top of #19460 pr/ipc-connect.54

ryanofsky committed:
interfaces, refactor: Change WalletLoader::restoreWallet parameter order

Libmultiprocess requires output parameters to be ordered after input parameter,
so move warnings parameter last. Problematic order was introduced in
4ec2d18a0734f44c0a74f05b59ad1269d323dfdb from
https://github.com/bitcoin-core/gui/pull/877

ryanofsky committed:
Merge branch 'pr/ipc' into pr/ipc-connect

ryanofsky committed:
Squashed 'src/ipc/libmultiprocess/' changes from 28e056576a3..393fce333e4

393fce333e4 proxy: fix BuildList to use non-const iteration for interface types
c958ec11d0c type-chrono: Add CustomBuildField/CustomReadField overloads for std::chrono::time_point
e8de5c7b68e Merge bitcoin-core/libmultiprocess#305: refactor: memcpy to std::ranges::copy to work around ubsan warn
9307e68e5aa Merge bitcoin-core/libmultiprocess#306: doc: Bump version 12 > 13
fac7b9b7f68 refactor: memcpy to std::ranges::copy to work around ubsan warn
1bd7025609e Merge bitcoin-core/libmultiprocess#297: test: add map serialization round-trip coverage
438fdd243d3 doc: Bump version 12 > 13
463d073cb8d test: rename vBool to vector_bool
85df2338456 test: add mapStringInt to foo.capnp to cover map serialization and deserialization

git-subtree-dir: src/ipc/libmultiprocess
git-subtree-split: 393fce333e4cf1fb49e67137fbe22589f2a0de57

ryanofsky committed:
Merge branch 'pr/ipc-connect' into pr/ipc-gui

ryanofsky committed:
Merge branch 'pr/ipc-chain' into pr/ipc

ryanofsky committed:
Merge commit '5635085e28c9b5365d622e8b0a1531df43561947' into HEAD
